CSVSource/src/test/java/io/vantiq/extsrc/CSVSource/TestCSVReader.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2020 Vantiq, Inc.
+ *
+ * All rights reserved.
+ * 
+ * SPDX: MIT
+ */
+
+package io.vantiq.extsrc.CSVSource;
+
+import static org.junit.Assert.assertTrue;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestCSVReader extends TestCSVBase{
+    
+    static Map<String,Object> config ;
+    static Map<String,Object> options ;
+    
+    @Before
+    public void setup() {
+        config = new HashMap<String,Object>(); 
+        
+        TestCSVConfig o = new TestCSVConfig();
+        config = o.minimalConfig();
+        options = o.createMinimalOptions();
+
+    }
+    @After
+    public void tearDown() {
+        List<CSVCore> sources = CSVMain.sources;
+        System.setSecurityManager(null);
+        if (sources != null) {
+            for (CSVCore s : sources) {
+                s.stop();
+            }
+            CSVMain.sources = null;
+        }
+    }
+    
+    @Test
+    public void testReadSimpleFile1Record() {
+
+        CreateFileForTest("a.csv","s,1,2");
+        ArrayList<Map<String,String>> content = CSVReader.execute("a.csv", config, null);
+        assertTrue("Unexpected lines of csv file in array", content.size() == 1);
+        assertTrue("Unexpected values in first of line of csv file in array", content.get(0).size() == 3);
+
+        String v = content.get(0).get("value");
+        assertTrue("Unexpected field value in first of line of csv file in array", v.equals("s"));
+        assertTrue("Unexpected field flag in first of line of csv file in array", content.get(0).get("flag").equals("2"));
+        assertTrue("Unexpected field YScale in first of line of csv file in array", content.get(0).get("YScale").equals("1"));
+
+
+    }
+    @Test
+    public void testToManyEventsForSingleSegment() {
+
+        CreateFileForTest("a.csv","s;1;2");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+        AppendFileForTest("a.csv","s1;11;21");
+
+        config.put("maxLinesInEvent",3);
+        config.put("delimiter",";");
+
+        ArrayList<Map<String,String>> content = CSVReader.execute("a.csv", config, null);
+        assertTrue("Unexpected lines of csv file in array", content.size() == 3);
```

Comment:
this tests that the first event is correct.  The remainder are not.  You need to test those as well.  This may require some restructuring fo `CSVReader.execute()`.  You are going to need to return an array of events. You can add some item in the config to say to run in test mode (or use the null `oClient` to do that, but you simply can't just return the first one.

The way the `CSVReader.execute()` is currently constructed, the next and all subsequent events will be of size 1, (or them may get increasingly big with the last one containing the entire file), and may have no new data in them because the lines buffer (the `file` variable) is never reset.  So this stuff won't really work, I don't think.  You need some real tests that exercise things.

Reply:
i tested that , looks much better now